lib/checks/aria/aria-allowed-attr-evaluate.js

```diff
@@ -27,31 +33,80 @@ import { getRole, allowedAttr, validateAttr } from '../../commons/aria';
  */
 function ariaAllowedAttrEvaluate(node, options, virtualNode) {
   const invalid = [];
-
+  const rowAttrInvalid = [];
   const role = getRole(virtualNode);
   const attrs = virtualNode.attrNames;
   let allowed = allowedAttr(role);
-
   // @deprecated: allowed attr options to pass more attrs.
   // configure the standards spec instead
   if (Array.isArray(options[role])) {
     allowed = uniqueArray(options[role].concat(allowed));
   }
 
+  let tableRole;
+  function validateRowAttrs(validValue) {
+    // check if the parent exists otherwise a TypeError will occur (virtual-nodes specifically)
+    if (virtualNode.parent && role === 'row' && validValue) {
+      // TODO: look into memoizing getRole and removing this local cache
+      let tableMap = cache.get('aria-allowed-attr-table');
+      if (!tableMap) {
+        tableMap = new WeakMap();
+        cache.set('aria-allowed-attr-table', tableMap);
+      }
+
+      const table = closest(
+        virtualNode,
+        'table, [role="treegrid"], [role="table"], [role="grid"]'
+      );
+      tableRole = tableMap.get(table);
```

Comment:
Role computation isn't all that expensive. Does caching it this actually make a difference?

Reply:
Since we're potentially doing role computation on the same table element, I figured it would be slow enough on a table with thousands of rows in the same table (like https://web.archive.org/web/20190613132353/https://giveawaylisting.com/). This is also why I wanted the TODO left in there as a note that we should look at memoizing the `getRole` function since we can call it for the same element multiple times, but I need to do a performance audit on it to find out.